src/coreclr/vm/jithelpers.cpp

```diff
@@ -5378,21 +5417,33 @@ EXTERN_C void JIT_ReversePInvokeEnter(ReversePInvokeFrame* frame)
     {
         frame->currentThread = thread;
 
+#ifdef PROFILING_SUPPORTED
+        if (CORProfilerTrackTransitions())
+        {
+            ProfilerUnmanagedToManagedTransitionMD(frame->pMD, COR_PRF_TRANSITION_CALL);
+        }
+#endif
+
         // Manually inline the fast path in Thread::DisablePreemptiveGC().
         thread->m_fPreemptiveGCDisabled.StoreWithoutBarrier(1);
-        if (g_TrapReturningThreads.LoadWithoutBarrier() == 0)
+        if (g_TrapReturningThreads.LoadWithoutBarrier() != 0)
         {
-            return;
+            JIT_ReversePInvokeEnterRare2(frame);
         }
-
-        JIT_ReversePInvokeEnterRare2(frame);
-        return;
+    }
+    else
+    {
+        JIT_ReversePInvokeEnterRare(frame);
     }
 
-    JIT_ReversePInvokeEnterRare(frame);
+#ifndef FEATURE_EH_FUNCLETS
+    frame->record.m_pEntryFrame = thread->GetFrame();
+    frame->record.m_ExReg.Handler = (PEXCEPTION_ROUTINE)FastNExportExceptHandler;
+    INSTALL_EXCEPTION_HANDLING_RECORD(&frame->record.m_ExReg);
+#endif
```

Comment:
The SEH chain after this call is (in reverse) FastNExportExceptHandler->COMPlusFrameHandler->native fp exception handler->...

I'm not sure if we still need to add the handler here now that we already have a COMPlusFrameHandler hooked in by the JIT.

Reply:
Do all tests pass without this?

Reply:
Based on some of the test failures, I was going to try removing this to see if they pass.

Reply:
After removing this, the libraries test failures are failing because they're not finding the right type of SEH frame. So it looks like in some cases we still need this (or we need to remove the assert that's firing).

Reply:
I am afraid we may actually need this to handle EH interop cases like this:

unmanaged code with `__try/__except` -> reverse P/Invoke -> managed code with try/catch -> P/Invoke -> unmanaged  that throws exception

I think we need to have a `fs:[0]` chain entry on reverse P/Invoke in this case. Otherwise, I do not see how the managed code with try/catch can catch the exception thrown by the unmanaged code in this case.

Reply:
I've reverted the change that removes adding the record locally and I've been debugging the crashes that change causes.

It seems like an invariant in the x86 exception handling code in the runtime that the managed SEH handler for a block of managed methods is not in the stack frame of the outermost managed method in the block.

In the code on master, the stack looks as follows:

| Master |
|---------|
| UnmanagedCallersOnly EBP Frame |
| NExport SEH record |
| emitted NDirect stub EBP frame |
| Native code EBP frame |
| Managed code EBP frame |


In this branch, we set up the NExport SEH record in the UnmanagedCallersOnly method's stack variables:

| Master |
|---------|
| NExport SEH record |
| UnmanagedCallersOnly EBP Frame |
| emitted NDirect stub EBP frame |
| Native code EBP frame |
| Managed code EBP frame |

As a result, when we unwind the stack frame for the UnmanagedCallersOnly method, we violate the invariant that the SEH record for a given frame is still on the stack.

I see a few solutions for this:

1.  Create an extra stack frame and copy all of the incoming stack args.
2. Somehow teach the JIT to emit a nested EBP frame just for the SEH record and adjust the incoming args stack offsets so we don't need to copy the args on the stack.
3. Teach the exception handling code in the runtime to handle the case where the SEH frame lives on the same stack frame of the outermost managed method in a block of managed methods instead of outside of the stack frame.


I have a prototype of option 3 that mainly just modifies asserts that were hit with the original NExport SEH record setup in this PR. I've tested it locally and it seems to work correctly. I'll push it out to see what CI thinks.

Reply:
/cc @janvorli

Reply:
I agree that option 3 is the best way to go.

For testing, you may also want to run the managed C++ test suite from C++ team. It should have better coverage of various corner cases that this can hit.

Reply:
I've validated that this PR passes the managed C++ test suite.